### PR TITLE
Implement multinode module

### PR DESCRIPTION
**Pull request by DamonLee5**
This PR contains the following updates in two files:
1. multinode.py: Core code to access a parallel cluster and distribute a function call across multiple nodes. 
2. demo_multinode_4D_shepp_logan.py: Demonstrate how to use the multinode to do multiple 3D reconstructions.

### CODE REVIEW

File: demo/demo_multinode_4D_shepp_logan.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+    ROR, boundary_size = mbircone.cone3D.compute_img_size(num_views, num_det_rows, num_det_channels,
+                                                          dist_source_detector,
+                                                          magnification,
+                                                          channel_offset=channel_offset, row_offset=row_offset,
+                                                          delta_pixel_detector=delta_pixel_detector,
+                                                          delta_pixel_image=delta_pixel_image)
+    Nz, Nx, Ny = ROR
+    img_slices_boundary_size, img_rows_boundary_size, img_cols_boundary_size = boundary_size
+    print('ROR of the recon is:', (Nz, Nx, Ny))
```
Comment (dyang37) [inline demo/demo_multinode_4D_shepp_logan.py:123]:
```
print('Region of reconstruction (ROR) shape is:', (Nz, Nx, Ny))
```

File: demo/demo_multinode_4D_shepp_logan.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+    # Set phantom parameters to generate a phantom inside ROI according to ROR and boundary_size.
+    # All valid pixels should be inside ROI.
+    num_rows_cols = Nx - 2 * img_rows_boundary_size  # Assumes a square image
+    num_slices_phantom = Nz - 2 * img_slices_boundary_size
+    print('ROI and shape of phantom is:', num_slices_phantom, num_rows_cols, num_rows_cols)
```
Comment (dyang37) [inline demo/demo_multinode_4D_shepp_logan.py:129]:
```
print('Region of interest (ROI) shape is: ', (num_slices_phantom, num_rows_cols, num_rows_cols), ' . This is also the shape of the generated phantom image volume at each time point.')
```

File: demo/demo_multinode_4D_shepp_logan.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    display_slice = img_slices_boundary_size + int(0.4 * num_slices_phantom)
+    display_x = num_rows_cols // 2
+    display_y = num_rows_cols // 2
+    display_view = 0
+
+    # Generate a 3D phantom
+    phantom = mbircone.phantom.gen_shepp_logan_3d(num_rows_cols, num_rows_cols, num_slices_phantom)
+    print('Generated phantom shape = ', np.shape(phantom))
+    phantom = mbircone.cone3D.pad_roi2ror(phantom, boundary_size)
+    print('Padded phantom shape = ', np.shape(phantom))
```
Comment (dyang37) [inline demo/demo_multinode_4D_shepp_logan.py:141]:
```
print('Padding phantom from ROI to ROR. Padded phantom shape = ', np.shape(phantom))
```

File: demo/demo_multinode_4D_shepp_logan.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+    # Set display indexes
+    display_slice = img_slices_boundary_size + int(0.4 * num_slices_phantom)
+    display_x = num_rows_cols // 2
+    display_y = num_rows_cols // 2
+    display_view = 0
+
+    # Generate a 3D phantom
+    phantom = mbircone.phantom.gen_shepp_logan_3d(num_rows_cols, num_rows_cols, num_slices_phantom)
+    print('Generated phantom shape = ', np.shape(phantom))
```
Comment (dyang37) [inline demo/demo_multinode_4D_shepp_logan.py:139]: print('Generating 3D phantom with same shape as ROI. Generated 3D phantom shape = ', np.shape(phantom))

File: demo/demo_multinode_4D_shepp_logan.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    phantom = mbircone.cone3D.pad_roi2ror(phantom, boundary_size)
+    print('Padded phantom shape = ', np.shape(phantom))
+    print()
+
+    # ###########################################################################
+    # Generate a 4D shepp logan phantom.
+    # ###########################################################################
+
+    print("Generating 4D simulated data by rotating the 3D shepp logan phantom by increasing degree per time point ...")
```
Comment (dyang37) [inline demo/demo_multinode_4D_shepp_logan.py:148]:
```
print("Generating 4D simulated data by rotating the 3D phantom in the positive angle direction ...")
```

**Review by dyang37 — APPROVED** [review]

**Comment by dyang37** [comment]
I tested this PR with a fresh conda env on both brown and Gilbreth. The multinode demo worked on both machines. Functionality-wise I think the PR is ready to be merged.

In the code review above I made some comments regarding the grammar and formatting of the print messages both in the demo script as well as scatter_gather() function. They are only suggestions and it is not necessary to adopt any/all of them to merge this PR.

As a side note (and Charlie also mentioned this last Thursday), I think it is necessary to write some additional instructions for the usage of specific machines like brown and Gilbreth, and keep it within our research lab. I consider myself to be relatively familiar with Gilbreth and Brown, yet it still took me two attempts on Gilbreth to type in the correct cluster arguments.

**Comment by cabouman** [comment]
I updated the print statements based on Diyu's suggestion. 
Greg, should I accept this PR?

**Comment by gbuzzard** [comment]
I think it's ok to accept, but I would like to see this by the end of the week:

additional instructions for the usage of specific machines like brown and Gilbreth, and keep it within our research lab. I consider myself to be relatively familiar with Gilbreth and Brown, yet it still took me two attempts on Gilbreth to type in the correct cluster arguments.

Greg
________________________________

**Comment by cabouman** [comment]
Wenrui and Diyu, please start a new branch with the documentation updates that Prof. Buzzard suggested.

**Comment by DamonLee5** [comment]
> Wenrui and Diyu, please start a new branch with the documentation updates that Prof. Buzzard suggested.

Sure, I will add additional instructions in the documentation.

**Pull request merged**